renovate/src/Renovate/BinaryFormat/ELF.hs

```diff
@@ -465,19 +486,128 @@ doRewrite cfg hdlAlloc loadedBinary symmap strat = do
+-- virtual address of the program header table (@AT_PHDR@), and places the
+-- result in the aux vector. This value is used by e.g. glibc's thread-local
+-- storage initialization code. The address it calculates is
+--
+-- @AT_PHDR = (elf_ppnt->p_vaddr - elf_ppnt->p_offset) + exec->e_phoff@
+--
+-- where elf_ppnt is the first LOAD segment appearing in the program header
+-- table (see linux/fs/binfmt_elf.c). QEMU's user-mode emulation also
+-- calculates a value for @AT_PHDR@, but it does so differently. It uses
+--
+-- @AT_PHDR = min_i(phdr[i].p_vaddr - phdr[i].p_offset) + exec->e_phoff@
+--
+-- where @min_i@ is the minimum over indices @i@ of LOAD segments. We must
+-- make these values coincide, so we have to ensure that the virtual address
+-- of the new PHDR segment minus its offset must be the minimal difference
+-- between any LOAD segment's address and offset.
+--
+-- See the relevant QEMU bug report for additional details:
+-- https://bugs.launchpad.net/qemu/+bug/1885332
 --
 -- NOTE: We probably want to think more carefully about the alignment of this
 -- address.  Currently, we page align it.
-phdrSegmentAddress :: Num a => a
-phdrSegmentAddress = 0x900000
+--
+-- This function assumes:
+--
+-- 1. There are no segments whose images overlap in the virtual address space
+-- 2. There is at least one LOAD segment
+-- 3. The size and file offset of the program header segment don't depend on
+--    the virtual address chosen
+choosePHDRSegmentAddress ::
+  (w ~ MM.ArchAddrWidth arch, E.ElfWidthConstraints w) =>
+  proxy arch ->
+  E.Elf w ->
+  ElfRewriter lm arch (E.ElfWordType w)
+choosePHDRSegmentAddress _proxy elf = do
+  let phdrs = E.allPhdrs (E.elfLayout elf)
+
+  -- A high (hopefully unused), page-aligned address
+  let defaultAddress = 0x900000
+
+  -- To figure out where to put this new segment, we'll need to know its offset
+  -- and how big it is, so we first append it at an arbitrary address and get
+  -- those values.
+  let fakePhdrSeg = phdrSegment (nextSegmentIndex elf) defaultAddress
+  ((), fakeELF) <- appendSegment fakePhdrSeg elf
+  let fakePhdrs = E.allPhdrs (E.elfLayout fakeELF)
+  fakePhdrSegment <-
+    case filter ((== E.PT_PHDR) . E.phdrSegmentType) fakePhdrs of
+      [seg] -> pure seg
+      phdrs' ->
+        fail $ "Internal error: Wrong number of PT_PHDR segments: " ++ show phdrs'
+  let requiredSize = E.phdrMemSize fakePhdrSegment
+  let E.FileOffset projectedOffset = E.phdrFileStart fakePhdrSegment
+
+  let segInfos = mapMaybe makeLoadSegmentInfo phdrs
+  case NEL.nonEmpty segInfos of
+    Nothing -> fail (unlines ("Internal error: No LOAD segments?" : map show phdrs))
+    Just segmentInfos ->
+      case findSpaceForPHDRs segmentInfos projectedOffset requiredSize of
+        Nothing ->
+          -- This is fine in practice if there's no thread-local storage.
+          if null $ filter ((== E.PT_TLS) . E.phdrSegmentType) phdrs
+          then pure defaultAddress
+          else fail $ unlines $
```

Comment:
We could also consider making this not fail if the caller explicitly turns on some option saying "I promise not to run this binary in QEMU user-mode".

Reply:
@travitch points out that it's hazardous to have separate code paths for binaries we run in QEMU user-mode and those we don't.

Reply:
I think we're leaving this as-is for now, "most" binaries seem to pass at this point.